src/storage/lock_manager/deadlock.rs

```diff
@@ -222,126 +230,42 @@ impl Inner {
 
     fn reset(&mut self) {
         self.leader_client.take();
-        self.recving.borrow_mut().clear();
-        self.failed.borrow_mut().clear();
         self.detect_table.borrow_mut().clear();
     }
 
-    fn watch_member_change(&mut self, handle: &Handle, pd_client: &Arc<RpcClient>) -> Result<()> {
-        let stores = pd_client.get_all_stores(true)?;
-        // The leader of region 1 is leader of deadlock detector.
-        let (_, leader) = pd_client.get_region_and_leader(&[0])?;
-        self.update_addrs_if_needed(stores);
-        self.update_leader_if_needed(&handle, leader);
-        self.handle_failed_requests_if_needed(&handle);
-        Ok(())
-    }
-
-    fn update_addrs_if_needed(&self, stores: Vec<metapb::Store>) {
-        self.addrs.replace(
-            stores
-                .into_iter()
-                .filter(|store| store.get_id() != self.store_id)
-                .map(|mut store| (store.get_id(), store.take_address()))
-                .collect(),
-        );
-    }
-
-    fn update_leader_if_needed(&mut self, handle: &Handle, leader: Option<metapb::Peer>) {
-        if let Some(leader) = leader {
-            let leader_id = leader.get_store_id();
-            if leader_id != self.leader_id {
-                if leader_id != self.store_id {
-                    self.become_follower(handle, leader_id);
-                } else {
-                    self.become_leader(handle);
-                }
+    fn change_role_if_needed(&mut self, leader_id: u64, leader_addr: String) {
+        match self.leader_info {
+            Some((id, ref addr)) if id == leader_id && *addr == leader_addr => {
+                debug!("leader not change"; "leader_id" => leader_id, "leader_addr" => leader_addr);
             }
-        }
-    }
-
-    fn handle_failed_requests_if_needed(&self, handle: &Handle) {
-        if self.is_leader() {
-            self.failed.borrow().iter().for_each(|id| {
-                let mut recving = self.recving.borrow_mut();
-                let new_client = self.reconnect_follower(*id);
-                if new_client.is_some() {
-                    let new_client = new_client.unwrap();
-                    self.get_wait_for_entries(handle, *id, &new_client);
-                    recving.insert(*id, new_client);
+            _ => {
+                if self.store_id != leader_id {
+                    self.become_follower(leader_id);
                 } else {
-                    info!("retry to get_wait_for_entries failed(address not found)"; "store_id" => *id);
-                    recving.remove(id);
+                    self.become_leader();
                 }
-            });
-            self.failed.borrow_mut().clear();
+                self.leader_info.replace((leader_id, leader_addr));
+            }
         }
     }
 
-    fn become_leader(&mut self, handle: &Handle) {
+    fn become_leader(&mut self) {
```

Comment:
The only difference between `become_leader` and `become_follower` is log message.
I think we can inline the two functions into `change_role_if_needed`.